htlcswitch/circuit.go

```diff
@@ -0,0 +1,158 @@
+//
+// NOTE: In current implementation of HTLC switch, the payment circuit might be
+// uniquely identified by payment hash but in future we implement the payment
+// fragmentation which makes possible for number of payments to have
+// identical payments hashes, but different source and destinations.
+//
+// For example if Alice(A) want to send 2BTC to Bob(B), then payment will be
+// split on two parts and node N3 will have circuit with the same payment hash,
+// and destination, but different source (N1,N2).
+//
+//	    	  1BTC    N1   1BTC
+//    	      + --------- o --------- +
+//      2BTC  |	                      |  2BTC
+// A o ------ o N0	           N3 o ------ o B
+//	      |		              |
+// 	      + --------- o --------- +
+//	         1BTC     N2   1BTC
+//
+type PaymentCircuit struct {
+	// PaymentHash used as uniq identifier of payment (not payment circuit)
+	PaymentHash CircuitKey
+
+	// Src is the channel id from which add HTLC request is came from and
+	// to which settle/cancel HTLC request will be returned back. Once the
+	// switch forwards the settle message to the source the circuit is
+	// considered to be completed.
+	Src *wire.OutPoint
+
+	// Dest is the channel id to which we propagate the HTLC add request
+	// and from which we are expecting to receive HTLC settle request back.
+	Dest *wire.OutPoint
+}
+
+// NewPaymentCircuit creates new payment circuit instance.
+func NewPaymentCircuit(src, dest *wire.OutPoint, key CircuitKey) *PaymentCircuit {
+	return &PaymentCircuit{
+		Src:         src,
+		Dest:        dest,
+		PaymentHash: key,
+	}
+}
+
+// ID returns unique id of payment circuit.
+func (a *PaymentCircuit) ID() string {
+	return a.Src.String() +
+		a.Dest.String() +
+		a.PaymentHash.String()
+}
+
+// isEqual checks the equality of two payment circuits.
+func (a *PaymentCircuit) IsEqual(b *PaymentCircuit) bool {
+	return bytes.Equal(a.PaymentHash[:], b.PaymentHash[:]) &&
+		a.Src.String() == b.Src.String() &&
+		a.Dest.String() == b.Dest.String()
+}
+
+// circuitMap is a thread safe, persistent storage of circuits. Each
+// circuit key (payment hash) might have numbers of circuits corresponding to it
+// because of future payment fragmentation, now every circuit might be uniquely
+// identified by payment hash (1-1 mapping).
+type circuitMap struct {
+	mutex    sync.RWMutex
+	circuits map[CircuitKey][]*PaymentCircuit
+}
+
+// newCircuitMap initialized circuit map with previously stored circuits and
+// return circuit map instance.
+func newCircuitMap() (*circuitMap, error) {
+	m := &circuitMap{
+		circuits: make(map[CircuitKey][]*PaymentCircuit),
+	}
+
+	return m, nil
+}
+
+// add function add circuit in circuit map, and also save it in database in
+// thread safe manner.
+func (m *circuitMap) add(circuit *PaymentCircuit) error {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	circuits, ok := m.circuits[circuit.PaymentHash]
+	if ok {
+		for _, c := range circuits {
+			if circuit.IsEqual(c) {
```

Comment:
Am I right that the only one case for such behaviour, where we have the circuit with the same  three-tuple is `debug` mode?